Draft - Pause menu

OSGH-web opened the pull request:
- Initial pause menu implementation
- In Game
  - Press `P` to pause/unpause
  - Press `A` to return to level select screen
  - Press `Q` to close the game window
- In Level select
  - Press `Q` to close the game window
 
<img width="640" alt="Screenshot 2025-04-07 at 10 51 36 PM" src="https://github.com/user-attachments/assets/a6619b7a-93fb-4f86-9b89-028d7dc0a1b3" />

- i'm leaving it as a draft because all the levels i made aren't actually instances of level_base.gd...

JakeTaussig commented on scripts/pause_menu.gd:
```diff
@@ -0,0 +1,48 @@
+		print("warning: parent is not pausable")
+	init_mute_audio_button()
+
+func _input(_event: InputEvent) -> void:
```
These work even while the pause menu is not brought up. Is this intended? I think they should only work when the pause menu is up

JakeTaussig commented:
I think It'd be nice if you could scroll through the pause menu with the arrow keys and press enter to select an option. Not the biggest deal though.

JakeTaussig commented:
I think it would be a good idea to combine this component with the one that appears after you've failed the level. Like, just bring up the pause menu and keep it there after the readyForRestart flag is true. Might require emitting a signal for the readyForRestart flag

JakeTaussig commented:
Sometimes it's easy to accidentally press a button more than once, triggering unintended behaviour. Maybe add a 0.1-0.2 second delay that prevents pressing another input option immediately.

OSGH-web commented:
> I think it would be a good idea to combine this component with the one that appears after you've failed the level. Like, just bring up the pause menu and keep it there after the readyForRestart flag is true. Might require emitting a signal for the readyForRestart flag

cool i was thinking the same thing. i'm gonna go through all the styles used in the UI components (end of level message, fuel meter, pause menu, level select screen) and make sure they all are controlled and styled in the same way. i'll probably also make it so you can't pause after you run out of fuel

> I think It'd be nice if you could scroll through the pause menu with the arrow keys and press enter to select an option. Not the biggest deal though.

makes a lot of sense bc you control the game + level selection with arrow keys and we'll probably eventually want controller support. i think this should be pretty easy to implement with focus settings on the buttons

JakeTaussig commented:
> > I think it would be a good idea to combine this component with the one that appears after you've failed the level. Like, just bring up the pause menu and keep it there after the readyForRestart flag is true. Might require emitting a signal for the readyForRestart flag
> 
> cool i was thinking the same thing. i'm gonna go through all the styles used in the UI components (end of level message, fuel meter, pause menu, level select screen) and make sure they all are controlled and styled in the same way. i'll probably also make it so you can't pause after you run out of fuel
> 
> > I think It'd be nice if you could scroll through the pause menu with the arrow keys and press enter to select an option. Not the biggest deal though.
> 
> makes a lot of sense bc you control the game + level selection with arrow keys and we'll probably eventually want controller support. i think this should be pretty easy to implement with focus settings on the buttons

Awesome, a couple last thoughts: Q is now mapped to both return_to_world_select() in the player and to exit the game in your new script. This means that even when the player loses, pressing Q still completely exits the game. Also, that option definitely shouldn't work outside of the pause menu imo to prevent accidentally quitting, especially considereing the q key is right next to W. I think what might be good is to have an option in the menu to exit the game, but no keybinding for it. People can always ALT+F4 if they want.

OSGH-web commented:
> Awesome, a couple last thoughts: Q is now mapped to both return_to_world_select() in the player and to exit the game in your new script. This means that even when the player loses, pressing Q still completely exits the game. Also, that option definitely shouldn't work outside of the pause menu imo to prevent accidentally quitting, especially considereing the q key is right next to W. I think what might be good is to have an option in the menu to exit the game, but no keybinding for it. People can always ALT+F4 if they want.

i agree, i removed Q quitting the game from the level select here https://github.com/OSGH-web/balloon-movement-demo/pull/17/commits/ba292aa39d430ee1d4f60bcb4f6de26a69d4a2f1

OSGH-web commented:
Updates:
- removed all keybindings except Escape to pause/unpause
  - 2e41682
  - c92e4c6
- added arrow key movement with A/Space/Enter to select a menu item
  - 010a698

OSGH-web commented:
I did a lot of UI cleanup
- created `scenes/UI/` and `scripts/UI/` -- 88b1359
- scaled up level select screen and updated styling to match pause menu and fuel label -- 9b94842
- move FuelLabel out of Player and into LevelBase to align with pause menu -- 1dbdab6f

We should prioritize creating an end of round screen but i think it should be a separate pr. it would have to be a separate scene but i can't think through how we would enable the screen.
